src/ddmd/func.d

```diff
@@ -3451,13 +3451,20 @@ extern (C++) class FuncDeclaration : Declaration
             {
                 Dsymbol st = tret.toDsymbol(null);
                 //printf("\t\treturning class/struct %s\n", tret.toChars());
-                for (Dsymbol s = st.parent; s; s = s.parent)
+
+                auto ad = st.isAggregateDeclaration();
+                assert(ad !is null);
+                // A closure is only needed when returning a _non-static_ struct/class.
+                if ((ad.storage_class & STCstatic) == 0)
                 {
-                    //printf("\t\t\tparent = %s %s\n", s.kind(), s.toChars());
-                    if (s == this)
+                    for (Dsymbol s = st.parent; s; s = s.parent)
                     {
-                        //printf("\t\treturning local %s\n", st.toChars());
-                        goto Lyes;
+                        //printf("\t\t\tparent = %s %s\n", s.kind(), s.toChars());
+                        if (s == this)
+                        {
+                            //printf("\t\treturning local %s\n", st.toChars());
+                            goto Lyes;
```

Comment:
It turns out it is hard to find a test case where this case is really needed. E.g.
```d
auto goo(int a, bool b) {
    int x = 2;

    struct SInside {
        int getX() { return x; }
    }

    SInside res;
    lazyfun(a);
    return res;
}
```
this will not trigger the codepath, because `requiresClosure` is already set at the start of `needsClosure()` resulting in an early return. Removing `getX` does trigger the codepath, but in that case we still actually don't need a closure (or is it strictly required by the spec?).